File: benchmarks/fixtures/webapp_go/internal/api/v1/payment.go
package v1

import (
    "webapp_go/internal/routes"
    "webapp_go/internal/validators"
    "webapp_go/pkg/logger"
)

var payV1Log = logger.GetLogger("api.v1.payment")

// HandlePayment handles v1 payment endpoint.
func HandlePayment(request map[string]interface{}) (map[string]interface{}, error) {
    payV1Log.Info("V1 HandlePayment")

    validator := validators.NewPaymentValidator()
    amount, _ := request["amount"].(string)
    currency, _ := request["currency"].(string)
    userID, _ := request["user_id"].(string)

    errs := validator.Validate(map[string]string{
        "amount": amount, "currency": currency, "user_id": userID,
    })
    if len(errs) > 0 {
        payV1Log.Error("Payment validation failed")
        return nil, errs[0]
    }

    result, err := routes.PaymentHandler(request)
    if err != nil {
        return nil, err
    }
    result["api_version"] = "v1"
    payV1Log.Info("V1 payment complete")
    return result, nil
}

File: benchmarks/fixtures/webapp_rs/validators/payment.rs
use crate::utils::helpers::get_logger;
use crate::validators::common::{validate_not_empty, validate_range};

/// Supported payment currencies.
const SUPPORTED_CURRENCIES: &[&str] = &["USD", "EUR", "GBP", "JPY", "CAD"];

/// Validate payment request parameters.
pub fn validate(amount: f64, currency: &str, source: &str) -> Result<(), Vec<String>> {
    let logger = get_logger("validators.payment");
    logger.info(&format!("Validating payment: {} {} from {}", amount, currency, source));
    let mut errors = Vec::new();
    if let Err(e) = validate_range("amount", amount, 0.01, 999999.0) {
        errors.push(e);
    }
    if !SUPPORTED_CURRENCIES.contains(&currency) {
        errors.push(format!("Unsupported currency: {}", currency));
    }
    if let Err(e) = validate_not_empty("source", source) {
        errors.push(e);
    }
    if errors.is_empty() {
        logger.info("Payment validation passed");
        Ok(())
    } else {
        logger.warn(&format!("Payment validation failed: {} errors", errors.len()));
        Err(errors)
    }
}

/// Validate refund request parameters.
pub fn validate_refund(transaction_id: &str, reason: &str) -> Result<(), Vec<String>> {
    let logger = get_logger("validators.payment");
    logger.info(&format!("Validating refund for txn: {}", transaction_id));
    let mut errors = Vec::new();
    if let Err(e) = validate_not_empty("transaction_id", transaction_id) {
        errors.push(e);
    }
    if let Err(e) = validate_not_empty("reason", reason) {
        errors.push(e);
    }
    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}
